validation/start.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/opencontainers/runtime-tools/specerror"
+	"github.com/opencontainers/runtime-tools/validation/util"
+	uuid "github.com/satori/go.uuid"
+)
+
+func wrap(t *tap.T, expected bool, specErr error, detailedErr error) {
+	t.Ok(expected, specErr.(*specerror.Error).Err.Err.Error())
+	diagnostic := map[string]string{
+		"reference": specErr.(*specerror.Error).Err.Reference,
+	}
+
+	if detailedErr != nil {
+		diagnostic["error"] = detailedErr.Error()
+		if e, ok := detailedErr.(*exec.ExitError); ok {
+			if len(e.Stderr) > 0 {
+				diagnostic["stderr"] = string(e.Stderr)
+			}
+		}
+	}
+	t.YAML(diagnostic)
+}
+
+func main() {
+	t := tap.New()
+	t.Header(0)
+
+	bundleDir, err := util.PrepareBundle()
+	if err != nil {
+		util.Fatal(err)
+	}
+	defer os.RemoveAll(bundleDir)
+
+	containerID := uuid.NewV4().String()
+
+	r, err := util.NewRuntime(util.RuntimeCommand, bundleDir)
+	if err != nil {
+		util.Fatal(err)
+	}
+	g := util.GetDefaultGenerator()
+	g.SetProcessArgs([]string{"sh", "-c", fmt.Sprintf("echo 'process called' >> %s", "/output")})
+	r.SetConfig(g)
+	output := filepath.Join(bundleDir, g.Spec().Root.Path, "output")
+
+	// start without id
+	err = r.Start()
+	wrap(t, err != nil, specerror.NewError(specerror.StartWithoutIDGenError, fmt.Errorf("start` operation MUST generate an error if it is not provided the container ID"), rspecs.Version), err)
+
+	// set id for the remaining tests
+	r.SetID(containerID)
+
+	// start a not `created` container - case one: non-exist container
+	err = r.Start()
+	wrap(t, err != nil, specerror.NewError(specerror.StartNotCreatedGenError, fmt.Errorf("attempting to `start` a container that is not `created` MUST generate an error"), rspecs.Version), err)
+
+	r.Create()
+	// start a `created` container
+	err = r.Start()
+	outputData, _ := ioutil.ReadFile(output)
+	// check the output
+	wrap(t, err == nil && string(outputData) == "process called\n", specerror.NewError(specerror.StartProcImplement, fmt.Errorf("`start` operation MUST run the user-specified program as specified by `process`"), rspecs.Version), err)
+
+	// start a not `created` container - case two: exist and `stopped`
+	util.WaitingForStatus(r, util.LifecycleStatusStopped, time.Second*10, time.Second*1)
+	err = r.Start()
+	// must generate an error
+	wrap(t, err != nil, specerror.NewError(specerror.StartNotCreatedGenError, fmt.Errorf("attempting to `start` a container that is not `created` MUST generate an error"), rspecs.Version), err)
+
+	outputData, _ = ioutil.ReadFile(output)
+	// must have no effect, it will not be something like 'process called\nprocess called\n'
+	wrap(t, string(outputData) == "process called\n", specerror.NewError(specerror.StartNotCreatedHaveNoEffect, fmt.Errorf("attempting to `start` a container that is not `created` MUST have no effect on the container"), rspecs.Version), err)
+
+	r.Delete()
+
+	g.Spec().Process = nil
+	r.SetConfig(g)
+	err = r.Create()
+	// 'create' could fail according to the spec when the process is nil
```

Comment:
Why “could fail”?  Is this the “[`create` can fail whenever it wants][1]” loophole (opencontainers/runtime-spec#813, opencontainers/runtime-spec#927)?  I don't see a need to call that out in this particular case.  If we want, we could adjust our `Create` wrapper to handle this in all cases.

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/runtime.md#L106

Reply:
If the process is nil, the config.json is invalid. It is up to the runtime to validate it in the 'create' since:  `The runtime MAY validate config.json against this spec.`

If a runtime validates config.json in 'create', it might return error. In this case, the container is not created, we can never test 'StartWithProcUnsetGenError'.  That is why I use 'Skip' in this case.

Reply:
> If the process is nil, the `config.json` is invalid.

That's not true since opencontainers/runtime-spec#701, which was part of runtime-spec v1.0.0-rc6 and later.

Reply:
go it.